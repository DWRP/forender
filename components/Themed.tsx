/**
 * Learn more about Light and Dark modes:
 * https://docs.expo.io/guides/color-schemes/
 */

import {
  Text as DefaultText,
  View as DefaultView,
  TextInput as DefaultTextnput,
  TouchableOpacity,
} from "react-native";

import Colors from "@/constants/Colors";
import { useColorScheme } from "../hooks/useColorScheme";
import { useState } from "react";
import FontAwesome from "@expo/vector-icons/FontAwesome";

type ThemeProps = {
  lightColor?: string;
  darkColor?: string;
};

export type TextProps = ThemeProps & DefaultText["props"];
export type ViewProps = ThemeProps & DefaultView["props"];
export type TextInputProps = ThemeProps & DefaultTextnput["props"];

export function useThemeColor(
  props: { light?: string; dark?: string },
  colorName: keyof typeof Colors.light & keyof typeof Colors.dark
) {
  const theme = useColorScheme() ?? "dark";
  const colorFromProps = props[theme];

  if (colorFromProps) {
    return colorFromProps;
  } else {
    return Colors[theme][colorName];
  }
}

export function Text(props: TextProps) {
  const { style, lightColor, darkColor, ...otherProps } = props;
  const color = useThemeColor({ light: lightColor, dark: darkColor }, "text");

  return <DefaultText style={[{ color }, style]} {...otherProps} />;
}

export function TextInput(props: TextInputProps) {
  const { style, lightColor, darkColor, secureTextEntry, ...otherProps } =
    props;
  const [hidePassword, setHidePassword] = useState(true);
  const color = useThemeColor({ light: lightColor, dark: darkColor }, "text");
  const borderColor = useThemeColor(
    { light: lightColor, dark: darkColor },
    "borderColor"
  );

  return (
    <View style={{ flexDirection: "row", alignItems: "center" }}>
      <DefaultTextnput
        secureTextEntry={secureTextEntry && hidePassword}
        placeholderTextColor={color}
        style={[
          {
            color,
            borderColor,
            borderWidth: 1,
            borderRadius: 8,
            width: "75%",
            marginVertical: 8,
            padding: 8,
          },
          style,
        ]}
        {...otherProps}
      />
      {secureTextEntry && (
        <TouchableOpacity onPress={() => setHidePassword(!hidePassword)}>
          <FontAwesome
            name={hidePassword ? "eye" : "eye-slash"}
            size={24}
            color={color}
            style={{ position: "absolute", left: -40, bottom: -10 }}
          />
        </TouchableOpacity>
      )}
    </View>
  );
}

export function View(props: ViewProps) {
  const { style, lightColor, darkColor, ...otherProps } = props;
  const backgroundColor = useThemeColor(
    { light: lightColor, dark: darkColor },
    "background"
  );

  return <DefaultView style={[{ backgroundColor }, style]} {...otherProps} />;
}
