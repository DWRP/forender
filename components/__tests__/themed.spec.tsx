import * as React from "react";
import renderer from "react-test-renderer";
import { Text, TextInput, View } from "../Themed"; // Substitua 'NomeDoArquivo' pelo nome do arquivo onde as funções estão definidas

describe("Testando a renderização dos componentes do tema", () => {
  it(`Text renderiza corretamente`, () => {
    const tree = renderer.create(<Text>Snapshot test for Text!</Text>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it(`TextInput renderiza corretamente`, () => {
    const tree = renderer
      .create(<TextInput placeholder="Snapshot test for TextInput!" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it(`TextInput(senha) renderiza corretamente`, () => {
    const tree = renderer
      .create(
        <TextInput
          placeholder="Snapshot test for Password TextInput!"
          secureTextEntry={true}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it(`View renderiza corretamente`, () => {
    const tree = renderer
      .create(<View style={{ width: 100, height: 100 }} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
