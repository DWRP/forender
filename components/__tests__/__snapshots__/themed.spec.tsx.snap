// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testando a renderização dos componentes do tema Text renderiza corretamente 1`] = `
<Text
  style={
    [
      {
        "color": "#000",
      },
      undefined,
    ]
  }
>
  Snapshot test for Text!
</Text>
`;

exports[`Testando a renderização dos componentes do tema TextInput renderiza corretamente 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#fff",
      },
      {
        "alignItems": "center",
        "flexDirection": "row",
      },
    ]
  }
>
  <TextInput
    placeholder="Snapshot test for TextInput!"
    placeholderTextColor="#000"
    style={
      [
        {
          "borderColor": "#000",
          "borderRadius": 8,
          "borderWidth": 1,
          "color": "#000",
          "marginVertical": 8,
          "padding": 8,
          "width": "75%",
        },
        undefined,
      ]
    }
  />
</View>
`;

exports[`Testando a renderização dos componentes do tema TextInput(senha) renderiza corretamente 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#fff",
      },
      {
        "alignItems": "center",
        "flexDirection": "row",
      },
    ]
  }
>
  <TextInput
    placeholder="Snapshot test for Password TextInput!"
    placeholderTextColor="#000"
    secureTextEntry={true}
    style={
      [
        {
          "borderColor": "#000",
          "borderRadius": 8,
          "borderWidth": 1,
          "color": "#000",
          "marginVertical": 8,
          "padding": 8,
          "width": "75%",
        },
        undefined,
      ]
    }
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <Text />
  </View>
</View>
`;

exports[`Testando a renderização dos componentes do tema View renderiza corretamente 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#fff",
      },
      {
        "height": 100,
        "width": 100,
      },
    ]
  }
/>
`;
